Proposing something to permit multiple segments and shapes

cmonagle opened the pull request:
# Supporting split service in GTFS-RT Trip Modifications

GTFS-RT Trip Modifications allow data-producers to communicate detours to active trips in real-time or near-real time. It is not, however, currently possible to tell data-consumers about service disruptions that result in service being operated in two or more discrete segments. This proposal includes a new, scoped extension to the ReplacementStop object, which will producers to define when through-service is not possible, and to associate disparate trip-segments with their respective shapes.

At a high level, when riders must disembark from a vehicle at a stop, that stop is considered "modified." As such, it should be removed from the selectedTrips, and may be added back as a replacement stop.

## Proposed change

To communicate split servive, data producers should add a replacement stop that contains the boolean "no_through_travel" at the last served stop. This may be a stop on the original itinerary, an existing stop that isn't typically served by this itinerary, or a temporary stop.

`no_through_travel` is conditionally required when `next_shape_id` is defined, but `no_through_travel` may be `true` without requiring a new shape (see example 4).

```protobuf
message TransitReplacementStopExtension {
  // Indicates that vehicles will not continue travelling. Riders must disembark
  // from the vehicle.
  optional bool no_through_travel = 1;

  // If the vehicle is not operating in revenue service beyond this stop, this field
  // must be populated with a new shape ID for the next segment of stops.bool
  // Additionally, when populated, this
  optional string next_shape_id = 2;
}

extend transit_realtime.ReplacementStop {
  optional TransitReplacementStopExtension transit_replacement_stop_extension = 9514;
}

```

## Example 1: Parade Disruption

A downtown parade prevents trips 100, 200, and 300 from following their usual route, splitting service into two segments:

- Riders must disembark at stop ID 4.
- Stops 4 through 7 are not served.
- Riders may board at stop 8.

```json
{
  "header": {
    "gtfsRealtimeVersion": "2.0",
    "timestamp": "1738697269"
  },
  "entity": [
    {
      "id": "shape_stops_1-4",
      "shape": { "shapeId": "shape_stops_1-4", "encodedPolyline": "[...]" }
    },
    {
      "id": "shape_stops_8-11",
      "shape": { "shapeId": "shape_stops_8-11", "encodedPolyline": "[...]" }
    },
    {
      "id": "modification_1",
      "tripModifications": {
        "selectedTrips": [
          { "tripIds": ["100", "200", "300"], "shapeId": "shape_stops_1-4" }
        ],
        "modifications": [
          {
            "startStopSelector": { "stopId": "4" },
            "endStopSelector": { "stopId": "7" },
            "replacementStops": [
              {
                "stop_id": "4",
                ".transit_replacement_stop_extension": {
                  "no_through_travel": true,
                  "next_shape_id": "shape_stops_8-11"
                }
              }
            ]
          }
        ]
      }
    }
  ]
}
```

![tm-split-service-example_Example 1](https://github.com/user-attachments/assets/597af59c-4ffb-4650-b45e-06e14fbf0083)

## Example 2: Bridge Closure

A temporary bridge closure forces trips 400, 500, and 600 to operate in two segments. To faciliate rider travel, it is detouring to the ferry slips on either bank.

- The first segment ends at a temporary stop (`ferry_slip_e`).
- Riders must disembark at `ferry_slip_e`.
- The shape for stops 1 to `ferry_slip_e` is defined in `selectedTrips`, while the shape from `ferry_slip_w` to stop 11 is defined in the temporary stop `ferry_slip_e`.

```json
{
  "header": { "gtfsRealtimeVersion": "2.0", "timestamp": "1738697269" },
  "entity": [
    {
      "id": "shape_stops_1-ferry_slip_east",
      "shape": {
        "shapeId": "shape_stops_1-ferry_slip_east",
        "encodedPolyline": "[...]"
      }
    },
    {
      "id": "shape_stops_ferry_slip_west-11",
      "shape": {
        "shapeId": "shape_stops_ferry_slip_west-11",
        "encodedPolyline": "[...]"
      }
    },
    {
      "id": "modification_1",
      "tripModifications": {
        "selectedTrips": [
          {
            "tripIds": ["400", "500", "600"],
            "shapeId": "shape_stops_1-ferry_slip_east"
          }
        ],
        "modifications": [
          {
            "startStopSelector": { "stopId": "5" },
            "endStopSelector": { "stopId": "7" },
            "replacementStops": [
              {
                "stop_id": "ferry_slip_east",
                ".transit_modification_extension": {
                  "no_through_travel": true,
                  "next_shape_id": "shape_stops_ferry_slip_west-11"
                }
              },
              {
                "stop_id": "ferry_slip_west"
              }
            ]
          }
        ]
      }
    }
  ]
}
```

![tm-split-service-example_Example 2](https://github.com/user-attachments/assets/664f8567-6afd-4ec8-8f1d-a8f414f12d48)

## Example 3: Flooded Streets

A water main break has complicated downtown travel. Several streets are impassible and service is divided into three distinct segments with separate shapes.

- Riders must disembark at stops 3 and 7.

```json
{
  "header": { "gtfsRealtimeVersion": "2.0", "timestamp": "1738697269" },
  "entity": [
    {
      "id": "shape_stops_1-3",
      "shape": { "shapeId": "shape_stops_1-3", "encodedPolyline": "[...]" }
    },
    {
      "id": "shape_stops_4-7",
      "shape": { "shapeId": "shape_stops_4-7", "encodedPolyline": "[...]" }
    },
    {
      "id": "shape_stops_9-11",
      "shape": { "shapeId": "shape_stops_9-11", "encodedPolyline": "[...]" }
    },
    {
      "id": "modification_1",
      "tripModifications": {
        "selectedTrips": [
          { "tripIds": ["700", "800", "900"], "shapeId": "shape_stops_1-3" }
        ],
        "modifications": [
          {
            "startStopSelector": { "stopId": "3" },
            "endStopSelector": { "stopId": "3" },
            "replacementStops": [
              {
                "stop_id": "3",
                ".transit_modification_extension": {
                  "no_through_travel": true,
                  "next_shape_id": "shape_stops_4-7"
                }
              }
            ]
          },
          {
            "startStopSelector": { "stopId": "7" },
            "endStopSelector": { "stopId": "8" },
            "replacementStops": [
              {
                "stop_id": "7",
                ".transit_modification_extension": {
                  "no_through_travel": true,
                  "next_shape_id": "shape_stops_9-11"
                }
              }
            ]
          }
        ]
      }
    }
  ]
}
```

![tm-split-service-example_Example 3](https://github.com/user-attachments/assets/4f20263f-3196-4c83-b737-c14c007717cf)

## Example 4: Shuttle bus bridge

There are emergency roadworks at an intersection. The maximum turning radius is significantly decreased, and the operator must use midi buses for the second half of a route. All stops are served, but riders must exit the articulated bus and board another vehicle.

Because stop 5 is effectively served twice (once dropping off by the articulated bus, and once by the midi-bus), we must include it twice in the replacement stops.

```json
{
  "header": { "gtfsRealtimeVersion": "2.0", "timestamp": "1738697269" },
  "entity": [
    {
      "id": "shape1",
      "shape": {
        "shapeId": "shape1",
        "encodedPolyline": "[...]"
      }
    },
    {
      "id": "modification_1",
      "tripModifications": {
        "selectedTrips": [
          {
            "tripIds": ["1000", "1100", "1200"],
            "shapeId": "shape1"
          }
        ],
        "modifications": [
          {
            "startStopSelector": { "stopId": "5" },
            "endStopSelector": { "stopId": "5" },
            "replacementStops": [
              {
                "stop_id": "5",
                ".transit_modification_extension": {
                  "no_through_travel": true
                }
              },
              {
                "stop_id": "5"
              }
            ]
          }
        ]
      }
    }
  ]
}
```
![tm-split-service-example_Example 4](https://github.com/user-attachments/assets/488bad93-41f0-4714-93e6-ebf5762eaf99)

# Communicating trip updates on split service segments

When service is split, we need to ensure that trip updates can unambiguously refer to a specific segment of the divided trip. To this end, we propose an extension to the [ModifiedTripSelector](https://gtfs.org/documentation/realtime/reference/#message-modifiedtripselector). Data producers must use the original trip ID in combination with the index of the split service. The value is one-base-indexed (starting at 1), and (unlike `sequence` values elsewhere in GTFS) must increment by one at a time. In example one,  `split_service_segment_index: 1` refers to the service between 1 and 4, while `split_service_segment_index: 2` refers to stops 8 through 9. Put differently, every time `no_through_travel` is true, the index increments by one.

As with all trip updates on modified trips, the trip_id must be included in the ModifiedTripSelector, but not the TripSelector. Additionally, providing a TripUpdate with a ModifiedTripSelector is the only way to create predictions on trips with split service.

```protobuf
message TransitModifiedTripSelector {
  optional int32 split_service_segment_index = 1;
}
```

cmonagle commented on transit-extensions.proto:
```diff
@@ -138,21 +138,20 @@ message TransitModificationExtension {
+
+  // This indicates that through travel is prohibited. Riders will be instructed by consuming apps to disembark from the vehicle.
+  // [If replacement_stops are empty] Through travel is prohibited between the start stop selector and end stop selecter
+  // [If replacement_stops are defined] Travel prohibited between the last replacement stop and the end stop selector.
```
- Illegal if the end stop selector is the last stop.

felixbouquet commented:
isn't it a bit strange to use `replacementStops` to notify service splits? I find the last example a bit confusing, for example.

Couldn't our extension simply be a separate field (for example : `service_splits`) with `stop_id` and `next_shape_id`? There'd be no need for an extra boolean, because I don't think this information is really necessary (if the field exists, we want to cut it, don't we?).

I also think that in the last example, it's still necessary to indicate the shapes, to avoid creating special cases and it makes the modification easier to process.

I think it would be clearer, the functions of each field would be well separated and respected and there would be no redundant information.

otherwise it's crystal clear!

cmonagle commented:
> isn't it a bit strange to use replacementStops to notify service splits? I find the last example a bit confusing, for example.
> Couldn't our extension simply be a separate field (for example : service_splits) with stop_id and next_shape_id? There'd be no need for an extra boolean, because I don't think this information is really necessary (if the field exists, we want to cut it, don't we?).

First, to confirm I'm understanding correctly: At the `modification` level, we'd have a new array (`service_splits`) which would include an array of stop selectors and shape_ids. 

There's one major issue I can identify:

-  stop_ids might be an ambiguous reference (loops, lollipops, etc.). In combination with stop_sequence, they could unambiguously reference a stop on the _original_ itinerary, but not the modified itinerary. To remove this ambiguity for modified itineraries, we'd need to introduce a new concept like "modified_stop_sequence." I think that's a can of worms and would add a lot of complexity.

Peter an I explored something similar (shapes_by_segment), defined at the trip_modification level, and ruled it out for a similar reason -- by definition, the itinerary is a moving target, and it would be onerous to impose a new, unambiguous identifier that applies to both the original and modified itineraries. 

This isn't what you're proposing, but for additional context, we also ruled out "one split per modification", defined at the modification level (no through service after the end stop selector).
- It would be impossible to define split service that occurs in the middle of a modification (see example 2).
- Because modifications cannot overlap (if they impact the same contiguous stops, they must be in the same modification), it becomes impossible to represent complicated combinations of splits and added stops.

> I also think that in the last example, it's still necessary to indicate the shapes, to avoid creating special cases and it makes the modification easier to process.

I'm open to requiring a new shape in either case if that makes it simpler. The cases where the shape is truly 100% identical are probably exceedingly rare.

felixbouquet commented:
Thanks for the summary of the reflection process, it helps me get up to date with you.
I understand your concerns about the ambiguity surrounding stop_id and stop_sequence, it's always the big struggle.

How about adding something like :
```
"splitted_trip" : {
	"number_of_new_stops_in_first_section" : uint
	"next_shape_id" : str
}
```
which can be at the modification level, so you can have several splits in a trip and add new stops unambiguously as you wish.
I don't think there's any loss of information compared to what you've suggested? and it can be clearer, we avoid using stop replacements if we don't need them.

with example 2:

```
    {
      "id": "modification_1",
      "tripModifications": {
        "selectedTrips": [
          {
            "tripIds": ["400", "500", "600"],
            "shapeId": "shape_stops_1-ferry_slip_east"
          }
        ],
        "modifications": [
          {
            "startStopSelector": { "stopId": "5" },
            "endStopSelector": { "stopId": "7" },
            "splitted_trip" :
            {
               "number_of_new_stops_in_first_section" : "1",
               "next_shape_id" : "shape_stops_ferry_slip_west-11"
             },
            "replacementStops": [
              {
                "stop_id": "ferry_slip_east",
              },
              {
                "stop_id": "ferry_slip_west"
              }
            ]
          }
        ]
      }
    }
```
          
with example 3:
```
    {
      "id": "modification_1",
      "tripModifications": {
        "selectedTrips": [
          { "tripIds": ["700", "800", "900"], "shapeId": "shape_stops_1-3" }
        ],
        "modifications": [
          {
            "startStopSelector": { "stopId": "3" },
            "endStopSelector": { "stopId": "3" },
            "splitted_trip" :
            {
               "number_of_new_stops_in_first_section" : "0",
               "next_shape_id" : "shape_stops_4-7"
             },
            "replacementStops": []
          },
          {
            "startStopSelector": { "stopId": "7" },
            "endStopSelector": { "stopId": "8" },
            "splitted_trip" :
            {
               "number_of_new_stops_in_first_section" : "0",
               "next_shape_id" : "shape_stops_9-11"
             },
            "replacementStops": []
          }
        ]
      }
    }
```

with example 5:
```
    {
      "id": "modification_1",
      "tripModifications": {
        "selectedTrips": [
          {
            "tripIds": ["1000", "1100", "1200"],
            "shapeId": "shape_stops_1-5"
          }
        ],
        "modifications": [
          {
            "startStopSelector": { "stopId": "5" },
            "splitted_trip" :
            {
               "number_of_new_stops_in_first_section" : "0",
               "next_shape_id" : "shape_stops_5-11"
             },
            "replacementStops": []
          }
        ]
      }
    }
```
(Note that in example 5, if we don't want the stop to be deleted, we simply don't set an `endStopSelector`, as indicated in [the current specification](https://gtfs.org/documentation/realtime/reference/#message-modification:~:text=If%20no%20stop_time%20is%20replaced%2C%20end_stop_selector%20must%20not%20be%20provided.%20It%27s%20otherwise%20required.))

> I'm open to requiring a new shape in either case if that makes it simpler.

yes, if it suits you I think that even if the shape doesn't change, you should always ask for one, this is not an optional field for me (especially since the shape is now cut in two).

cmonagle commented:
Thanks for the thoughtful reply!

I think this idea is workable.

### Naming

- Nit: `splitted_trip` should be `trip_split` or something similar (the past tense of split is just split, but here I think we'd want the noun anyway).
- `number_of_new_stops_in_first_section` is not the best name, because there may be multiple modifications with new stops that will end up being in the "first" section.
- There may be multiple sections within a single modification, so the cardinality is misleading anyway.
- It looks like we're really adding a new type of stop selector, the "replacement stop index", where 0 is the startStopSelector. This is an elegant solution to the stop-selector problem.
- maybe... "replacement_stop_index_selector_[where]_through_travel_prohibited" or "replacement_stop_index_of_disembarkation" (these are both too long but something along these lines).

### Ambiguities

At a higher level, I question whether it's really simpler to add an entirely different modification types. TM is based on a deceptively simple but powerful model: you just remove stops, and then add the replacement ones. Additionally, I think there's some subtle differences between your updated examples and the first document:

- For example three: your example effectively removes service at stops 3 and 7. They *are* served, but as dropoff-only. I guess we'd need to add back the stop? Unless we're changing the behaviour of the stop selectors (ie: if index zero is a split trip, it should not remove that stop), which would be inconsistent/confusing.
- I don't think example five is communicating the same thing as the first doc: Riders must disembark at replacement stop index 0 (stop 5). Then... do they board at stop 5 or the 6? We'd still need a replacement stop here.

As a nit, we'd also need to scope the `trip_split`s to a transit extension, but that doesn't meaningfully change anything.

felixbouquet commented:
- I'm completely flexible on names, `trip_split` works for me and I agree that `number_of_new_stops_in_first_section` is a confusing name this was just an example, so we can change the name without any problem for example into `number_of_new_stops_in_next_split` (which you'll note means the opposite, but perhaps more clearly...) to match the `next_shape_id`

- All your comments on drop_off stops are very accurate...in fact, the ideal would be to have an exclusive rather than inclusive range for stop_selectors when it's a modification with a split_service, but it's true that it can be really confusing...maybe we should create a new modification type with more appropriate fields yes...

cmonagle commented:
Re naming: I think we can find something that works. I'll put some thought into it.

Re: drop-off and "the ideal would be to have an exclusive rather than inclusive range for stop_selectors when it's a modification with a split_service"  I feel like we're straying a bit too far from the existing specification, but we could alternatively just add the stop back as a replacement stop?

felixbouquet commented:
> I'm open to requiring a new shape in either case if that makes it simpler. The cases where the shape is truly 100% identical are probably exceedingly rare.

We didn't continue this discussion so I'm not sure if we've decided this yet, but have we decided to make the `next_shape_id` conditionally required if `no_through_travel` is set?

I still don't like so much the missing shape in example 4, I think `next_shape_id` should be mandatory in this case just like `shapeId` is mandatory even if the shape doesn't change in the detour, otherwise it adds processing and unnecessary checks and I prefer to keep it simple (and it also differs from the original standard)

(sorry, I had forgotten this point, but when I implemented the changes it came back to me)
